src/jit/flowgraph.cpp

```diff
@@ -8241,90 +8241,381 @@ void Compiler::fgAddInternal()
         lvaTable[lvaSecurityObject].lvType = TYP_REF;
     }
 
-    /* Assume we will generate a single shared return sequence */
+    // Merge return points if required or beneficial
+    CLANG_FORMAT_COMMENT_ANCHOR;
 
-    ULONG returnWeight = 0;
-    bool  oneReturn;
-    bool  allProfWeight;
+#ifdef JIT32_GCENCODER
 
+    // X86 GC encoding has a hard limit of SET_EPILOGCNT_MAX epilogs.
+    const unsigned returnCountHardLimit = SET_EPILOGCNT_MAX;
+#else  // JIT32_GCENCODER
+
+    // We currently apply a hard limit of '4' to all other targets (see
+    // the other uses of SET_EPILOGCNT_MAX), though it would be good
+    // to revisit that decision based on CQ analysis.
+    const unsigned returnCountHardLimit = 4;
+#endif // JIT32_GCENCODER
+
+    // Define a helper class for merging return blocks (which we do when the input has
+    // more than the limit for this configuration).
     //
-    //  We will generate just one epilog (return block)
-    //   when we are asked to generate enter/leave callbacks
-    //   or for methods with PInvoke
-    //   or for methods calling into unmanaged code
-    //   or for synchronized methods.
-    //
-    if (compIsProfilerHookNeeded() || (info.compCallUnmanaged != 0) || opts.IsReversePInvoke() ||
-        ((info.compFlags & CORINFO_FLG_SYNCH) != 0))
+    // Notes: sets fgReturnCount, genReturnBB, and genReturnLocal.
+    class MergedReturns
     {
-        // We will generate only one return block
-        // We will transform the BBJ_RETURN blocks
-        //  into jumps to the one return block
-        //
-        oneReturn     = true;
-        allProfWeight = false;
-    }
-    else
-    {
-        //
-        // We are allowed to have multiple individual exits
-        // However we can still decide to have a single return
-        //
-        oneReturn     = false;
-        allProfWeight = true;
+        Compiler* comp;
 
-        // Count the BBJ_RETURN blocks and set the returnWeight to the
-        // sum of all these blocks.
-        //
-        fgReturnCount = 0;
-        for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+        // As we discover returns, we'll record them in `returnBlocks`, until
+        // the limit is reached, at which point we'll keep track of the merged
+        // return blocks in `returnBlocks`.
+        BasicBlock* returnBlocks[returnCountHardLimit];
+
+        // Each constant value returned gets its own merged return block that
```

Comment:
If this runs after the ValueNumbering phase I would suggest using the ValueNumber as the key instead of a constant.
That way a Select function could return argA or argB in several places in the method and those returns could be combined as with constants.

Reply:
This runs in `fgAddInternal`, well before value numbering.